llvm/include/llvm/Analysis/CachedBitsValue.h

```diff
@@ -0,0 +1,188 @@
+  ImplCachedBitsValue(const ImplCachedBitsValue &) = default;
+  ImplCachedBitsValue &operator=(const ImplCachedBitsValue &) = default;
+
+  ~ImplCachedBitsValue() = default;
+
+  [[nodiscard]] ValuePointerType getValue() { return Pointer; }
+  [[nodiscard]] ValuePointerType getValue() const { return Pointer; }
+
+  [[nodiscard]] const KnownBits &getKnownBits(unsigned Depth,
+                                              const SimplifyQuery &Q) const {
+    if (!hasKnownBits())
+      calculateKnownBits(Depth, Q);
+    return Known;
+  }
+
+  [[nodiscard]] KnownBits &getKnownBits(unsigned Depth,
+                                        const SimplifyQuery &Q) {
+    if (!hasKnownBits())
+      calculateKnownBits(Depth, Q);
+    return Known;
+  }
+
+  [[nodiscard]] const KnownBits &getKnownBits(const APInt &DemandedElts,
+                                              unsigned Depth,
+                                              const SimplifyQuery &Q) const {
+    if (!hasKnownBits())
+      calculateKnownBits(DemandedElts, Depth, Q);
+    return Known;
+  }
+
+  [[nodiscard]] KnownBits &getKnownBits(const APInt &DemandedElts,
+                                        unsigned Depth,
+                                        const SimplifyQuery &Q) {
+    if (!hasKnownBits())
+      calculateKnownBits(DemandedElts, Depth, Q);
+    return Known;
+  }
+
+  [[nodiscard]] bool hasKnownBits() const { return HasKnownBits; }
+
+  operator ValuePointerType() { return Pointer; }
+  ValuePointerType operator->() { return Pointer; }
+  ValueReferenceType operator*() { return *Pointer; }
+
+  operator ValuePointerType() const { return Pointer; }
+  ValuePointerType operator->() const { return Pointer; }
+  ValueReferenceType operator*() const { return *Pointer; }
+};
+} // namespace detail
+
+class CachedBitsConstValue : public detail::ImplCachedBitsValue<true> {
+public:
+  CachedBitsConstValue() = default;
+  CachedBitsConstValue(ValuePointerType Pointer)
+      : ImplCachedBitsValue(Pointer) {}
+  CachedBitsConstValue(Value *Pointer) : ImplCachedBitsValue(Pointer) {}
+  CachedBitsConstValue(ValuePointerType Pointer, const KnownBits &Known)
+      : ImplCachedBitsValue(Pointer, Known) {}
+
+  template <typename T, std::enable_if_t<ValuePointerConvertible<T>, int> = 0>
+  CachedBitsConstValue(const T &Value) : ImplCachedBitsValue(Value) {}
+
+  template <typename T, std::enable_if_t<ValuePointerConvertible<T>, int> = 0>
+  CachedBitsConstValue(const T &Value, const KnownBits &Known)
+      : ImplCachedBitsValue(Value, Known) {}
+
+  CachedBitsConstValue(CachedBitsConstValue &&) = default;
+  CachedBitsConstValue &operator=(CachedBitsConstValue &&) = default;
+  CachedBitsConstValue(const CachedBitsConstValue &) = default;
+  CachedBitsConstValue &operator=(const CachedBitsConstValue &) = default;
+
+  ~CachedBitsConstValue() = default;
+};
+
+class CachedBitsNonConstValue : public detail::ImplCachedBitsValue<false> {
```

Comment:
It looks like only CachedBitsConstValue is actually used -- do we really need CachedBitsNonConstValue and all the extra complexity that comes with it?

I feel like even if we need it somewhere in the future, we might be better off just adding a `const_cast` than jump through these hoops.

Reply:
I had originally made this class for using it in `simplifyICmpInst` because that uses a non-const pointer. Personally, I feel having an extra class is better than using `const_cast`, as it is a huge code smell for me.

Reply:
Would it be possible to make the class `CachedBitsValue<const Value *>` and `CachedBitsValue<Value *>` instead? That would also allow using variants like `CachedBitsValue<Instruction *>` and hopefully make the implementation simpler by dropping the detail class?

(Maybe in that case the CachedBitsValue name is no longer ideal and the `WithCache<Value *>` starts making more sense.)

Reply:
Sure! I'll look into it.